File: epi/PrimitiveMultiply.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class PrimitiveMultiply {
  @EpiTest(testDataFile = "primitive_multiply.tsv")
  public static long multiply(long x, long y) {
//    if(x == 0 || y == 0) {
//      return 0;
//    }
//
//    long rt = 0;
//
//    while(x != 0) {
//      if((x & 1) != 0) {
//        rt = add(rt, y);
//      }
//
//      x >>>= 1;
//      y <<= 1;
//    }
//
//    return rt;
    return epi.kt.PrimitiveMultiply.INSTANCE.multiply(x, y);
  }

  private static long add(long x, long y) {
    long k = 1, sum = 0;
    long carryIn = 0;
    long tX = x,tY = y;

    while(tX > 0 || tY > 0) {
      int kX = (int) (x & k), kY = (int) (y & k);
      long carryOut = (kX & kY) | (kX & carryIn) | (kY & carryIn);
      sum |= (kX ^ kY ^ carryIn);
      carryIn = carryOut << 1;
      k <<= 1;
      tX >>>= 1;
      tY >>>= 1;
    }

    return sum | carryIn;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PrimitiveMultiply.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/PrimitiveMultiply.kt
package epi.kt

object PrimitiveMultiply {
    fun multiply(x: Long, y: Long): Long {
        var lX = x
        var lY = y
        var rt = 0L

        while(lX > 0) {
            if(lX and 1 != 0L) {
                rt = sum(rt, lY)
            }

            lX = lX shr (1)
            lY = lY shl (1)
        }

        return rt
    }

    private fun sum(x: Long, y: Long): Long {
        var tX = x
        var tY = y
        var k = 1L
        var sum = 0L
        var carryIn = 0L

        while(tX > 0 || tY > 0) {
            val kX = x and k
            val kY = y and k
            val carryOut = (kX and kY) or (kX and carryIn) or (kY and carryIn)
            sum = sum or (kX xor(kY) xor(carryIn))
            carryIn = carryOut shl(1)
            k = k shl(1)
            tX = tX shr(1)
            tY = tY shr(1)
        }

        return sum or carryIn
    }
}